### PR TITLE
Changes done after testing vol 1

### PR DESCRIPTION
**Pull request by hvsio**
### 16/11/2020

**What I've done so far:**
1. 1 game, 1 user, all challenges completion - **passed**
2. 1 game, 2+ users, all challenges completion - **passed**, fixed resubmitting challenges if already completed
3. +20 submission reward - **not passed** and I could not map the place where the issue originates from, console.logs were printing the translated id but in the repo it was still the gedil id and resulted in errors during rewards assignment
4. Check CHALLENGE_OPENED job - **done**, I will implement the complete section of CHALLENGE_*** hooks in the next commit

**What I've noticed/changed:**
- no metrics in submissions not updated in Moshaak (not sure if its an issue or just a detail that we do not need for now)
- submission language not updated in Moshaak (not sure if its an issue or just a detail that we do not need for now)
- fixed resubmitting (could result in reopening of rewarded/unlocked challenge) (submission.processor.ts)
- second reward (20+ submissions) is not being assigned

### 
17/11/2020
**What I've done so far:**
1. 3 games, 1 user, all challenges completion (pay attention to the hooks) - **passed**
2. The same user enrolled in 3 different games (check rewards assignments, ids of game, player and submissions) - **passed**
3. Check the same user in 3 games, if they do not overlap somehow or mix - **passed**
4. 2 users enrolled in 3 diff games (check rewards assignments, ids of game, player and submissions)  - **passed**

**What I've noticed/changed:**
- Locked challenge D could have been completed without unlocking it first (therefore additional `challenge` in EvaluationArgs in `evaluate` Mutation), should secure it also for FAILED, REJECTED?
- @Mutation(() => SubmissionDto, { **nullable: true** }) evaluate - because if the challenge is locked/hidden we return null, without {nullable: true} it was complaining...
- Implemented CHALLENGE_*** jobs
-  `(node:332) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added to [Redis]. Use emitter.setMaxListeners() to increase limit` - error appearing in the console after creating all CHALLENGE_*** processors, I implemented EventEmitter.defaultMaxListeners as proposed by node.js but idk if there is a prettier way of doing it for NestJS, feel free to change it.
- eslint fixes

### CODE REVIEW

**Comment by josepaiva94** [comment]
1. Good.
2. Explain, please.
3. Fill an issue, please.
4. Thanks.

About notices, have you delete Mooshak docker image and pulled it again?

**Comment by hvsio** [comment]
2.  it's more of a personal note, at the point of testing multiple users completing challenges I've noticed that after sending a valid submission to a completed challenge results in assignment of the reward for the second time, it refers to the 3rd bullet: `fixed resubmitting (could result in reopening of rewarded/unlocked challenge) (submission.processor.ts)`

**Comment by josepaiva94** [comment]
So, this change was done `{ $or: [{ sourceId: { $eq: exerciseId } }, { sourceId: { $eq: null } }] },`?

I'm just confused about what you mean by "resubmitting"?

**Comment by hvsio** [comment]
its in challenge-status.service.ts 
```
async markAsCompleted(gameId: string, challengeId: string, playerId: string, date: Date): Promise<ChallengeStatus> {
    const temp: ChallengeStatus = await this.findByChallengeIdAndPlayerId(challengeId, playerId);	    const temp: ChallengeStatus = await this.findByChallengeIdAndPlayerId(challengeId, playerId);
    **if (temp.state === StateEnum.COMPLETED) {
      console.log('challenge already completed');
      return;
    }**
    const result: ChallengeStatus = await this.patch(temp.id, { state: StateEnum.COMPLETED, endedAt: date });	    const result: ChallengeStatus = await this.patch(temp.id, { state: StateEnum.COMPLETED, endedAt: date });
```

I've tested scenario where A,B,C were successfully completed and therefore unlocked D. Then I've submitted D successfully as well -> D is of state COMPLETED. But if you then submit fx. A again and it's successful -> the state of D changes from COMPLETED back to OPENED.

**Comment by hvsio** [comment]
`{ $or: [{ sourceId: { $eq: exerciseId } }, { sourceId: { $eq: null } }] }` - this was done while having the "20+ submissions" hook, when we save it to repo it does not have sourceId as it is a pretty general hook, not attached to any specific challenge or else
image

File: src/submission/submission.service.ts [hunk 1]
```diff
@@ -32,22 +35,34 @@ export class SubmissionService extends BaseService<Submission> {
...
-  async sendSubmission(gameId: string, exerciseId: string, playerId: string, file: IFile): Promise<Submission> {
-    const submission: Submission = await super.create({
-      game: gameId,
-      player: playerId,
-      exerciseId: exerciseId,
-    } as Submission);
+  async sendSubmission(
+    gameId: string,
+    exerciseId: string,
+    challengeId: string,
```
Comment (josepaiva94) [inline src/submission/submission.service.ts]: Problem is that an exercise can belong to many challenges. Why add challenge here?

Comment (hvsio) [reply src/submission/submission.service.ts]: To query the ChallengeStatus by Player_id and Challenge_id, if the challenge is locked or hidden then we should not create a submission imo

Comment (josepaiva94) [reply src/submission/submission.service.ts]: An exercise can belong to many challenges, that's why `sendSubmission` did not receive a `challengeId`. I suggest to process submission, but only complete a challenge if it's in OPENED or AVAILABLE state.

**Pull request merged**